File: php/bin/canon_check.php
<?php
// Reads the shared corpus (spec/corpus/canon.json) and confirms this PHP
// canonicalizer reproduces each vector's expected bytes. Agreement with the
// other implementations means a certificate signed by any of them verifies
// under this one. Invoked by interop/check-canon-php.sh; exits non-zero on any
// mismatch.
require __DIR__ . '/../src/autoload.php';

use Bonemesh\Canon;

$path = $argv[1] ?? null;
if ($path === null) {
    fwrite(STDERR, "usage: canon_check <canon.json>\n");
    exit(2);
}
$doc = json_decode(file_get_contents($path), true);
$failures = 0;
foreach ($doc['vectors'] as $v) {
    try {
        $got = Canon::canonicalize($v['cert']);
    } catch (\Throwable $e) {
        echo "FAIL {$v['name']}\n  error: {$e->getMessage()}\n";
        $failures++;
        continue;
    }
    if ($got === $v['canonical']) {
        echo "PASS {$v['name']}\n";
    } else {
        echo "FAIL {$v['name']}\n  got:  $got\n  want: {$v['canonical']}\n";
        $failures++;
    }
}
if ($failures > 0) {
    fwrite(STDERR, "$failures vector(s) mismatched\n");
    exit(1);
}
echo 'all ' . count($doc['vectors']) . " canon vectors match\n";

File: python/bin/canon_check.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = ["bonemesh"]
#
# [tool.uv.sources]
# bonemesh = { path = "..", editable = true }
# ///
"""Reads the shared canon corpus (spec/corpus/canon.json) and confirms this
implementation reproduces every canonical byte string exactly.

Agreement means a certificate signed by one implementation verifies under any
other. Invoked by interop/check-canon-python.sh.
"""

import json
import sys

from bonemesh.canon import canonicalize


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: canon_check <canon.json>", file=sys.stderr)
        return 2
    with open(sys.argv[1], encoding="utf-8") as fh:
        doc = json.load(fh)
    vectors = doc.get("vectors") or []
    if not vectors:
        print("corpus has no canon vectors", file=sys.stderr)
        return 1
    failures = 0
    for v in vectors:
        want = v["canonical"]
        got = canonicalize(v["cert"]).decode("utf-8")
        if got == want:
            print(f"PASS {v['name']}")
        else:
            print(f"FAIL {v['name']}\n  got:  {got}\n  want: {want}")
            failures += 1
    if failures:
        print(f"{failures} vector(s) mismatched", file=sys.stderr)
        return 1
    print("canonicalization reproduces every shared vector")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
